<a href="https://colab.research.google.com/github/keisyashakila/MSCI-151/blob/main/CW_4_KEISYA_.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
!pip install openpyxl

In [8]:
import pandas as pd

excel_file_path = '/MSCI151 Data new.xlsx'
df = pd.read_excel(excel_file_path)
print(df.head())

   id  Profitability   debt     RnD  Firm Age  Firm Size  CEO Experience  \
0   1           0.02  23.76   53.09      3.43      21.22               2   
1   2           3.99  41.85   45.30      3.22      21.83               4   
2   3          41.98  61.84   51.16      3.18      20.54               3   
3   4           8.48  25.61  115.49      3.26      20.59               3   
4   5          -1.94  74.24   13.57      3.26      18.73               3   

   foreign sales  CEO education  CEO Gender  marketing  product sentiment  \
0           9.16              2           0   0.801335           0.284925   
1          41.42              1           0   0.385019           0.579495   
2           5.88              3           1   0.839350           0.619687   
3          11.08              2           0   0.643304           0.048441   
4          24.61              1           0   0.747535           0.378619   

   politician on board  POB catagorical  
0             0.158781                

## Split the data

### Subtask:
Split the data into training and testing sets using an 80/20 ratio.

**Reasoning**:
Import the necessary function for splitting the data and then split the data into training and testing sets.

In [10]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

## Prepare the data

### Subtask:
Define the features (X) and target (y) variables.

**Reasoning**:
Define the features and target variables as specified in the instructions.

In [9]:
features = ['debt', 'RnD', 'Firm Age', 'Firm Size', 'CEO Experience', 'foreign sales', 'CEO education', 'CEO Gender']
X = df[features]
y = df['Profitability']

# Task
Analyze the dataset located at "/MSCI151 Data.xlsx". Print the column headings. Build a predictive model using the Random Forest algorithm to predict 'Profitability' (Y) based on the following features (X): 'debt', 'RnD', 'Firm Age', 'Firm Size', 'CEO Experience', 'foreign sales', 'CEO education', and 'CEO Gender'. Split the data into training and testing sets using an 80/20 ratio.

## Make predictions

### Subtask:
Use the trained model to make predictions on the testing data.

**Reasoning**:
Use the trained Random Forest Regressor model to predict the profitability on the test data.

In [13]:
y_pred = model.predict(X_test)

## Train the random forest model

### Subtask:
Train a Random Forest Regressor model on the training data.

**Reasoning**:
Import the Random Forest Regressor and train the model on the training data.

In [12]:
from sklearn.ensemble import RandomForestRegressor

model = RandomForestRegressor(random_state=42)
model.fit(X_train, y_train)

RandomForestRegressor(random_state=42)

## Prepare the data

### Subtask:
Define the features (X) and target (y) variables.


**Reasoning**:
Define the features and target variables as specified in the instructions.



In [11]:
features = ['debt', 'RnD', 'Firm Age', 'Firm Size', 'CEO Experience', 'foreign sales', 'CEO education', 'CEO Gender', 'marketing', 'product sentiment', 'POB catagorical']
X = df[features]
y = df['Profitability']

## Split the data

### Subtask:
Split the data into training and testing sets using an 80/20 ratio.


**Reasoning**:
Import the necessary function for splitting the data and then split the data into training and testing sets.



In [4]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

## Train the random forest model

### Subtask:
Train a Random Forest Regressor model on the training data.


**Reasoning**:
Import the Random Forest Regressor and train the model on the training data.



In [5]:
from sklearn.ensemble import RandomForestRegressor

model = RandomForestRegressor(random_state=42)
model.fit(X_train, y_train)

RandomForestRegressor(random_state=42)

## Make predictions

### Subtask:
Use the trained model to make predictions on the testing data.


**Reasoning**:
Use the trained Random Forest Regressor model to predict the profitability on the test data.



In [6]:
y_pred = model.predict(X_test)

## Evaluate the model

### Subtask:
Evaluate the performance of the model using appropriate metrics.


**Reasoning**:
Import necessary metrics and calculate the MSE and R2 scores.



In [7]:
from sklearn.metrics import mean_squared_error, r2_score

mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print(f"Mean Squared Error (MSE): {mse}")
print(f"R-squared (R2) Score: {r2}")

Mean Squared Error (MSE): 113.67692451119993
R-squared (R2) Score: -1.4687530565153741


## Summary:

### Data Analysis Key Findings

*   The dataset was successfully loaded and the specified features ('debt', 'RnD', 'Firm Age', 'Firm Size', 'CEO Experience', 'foreign sales', 'CEO education', and 'CEO Gender') and the target variable ('Profitability') were identified.
*   The data was split into training (80%) and testing (20%) sets.
*   A Random Forest Regressor model was trained on the training data.
*   Predictions were made on the testing data.
*   The model's performance was evaluated, resulting in a Mean Squared Error (MSE) of approximately 113.68 and an R-squared (R2) score of approximately -1.47.

### Insights or Next Steps

*   The negative R2 score indicates that the current model does not perform better than simply predicting the mean of 'Profitability'.
*   Further investigation is needed to understand why the model performed poorly, which could involve exploring different features, hyperparameter tuning of the Random Forest model, or considering alternative regression algorithms.


## Feature Importance

### Subtask:
Display the feature importance scores from the trained Random Forest model.

**Reasoning**:
Access the `feature_importances_` attribute of the trained Random Forest model and display the scores along with the corresponding feature names.

In [16]:
import pandas as pd

excel_file_path = '/MSCI151 Data new.xlsx'
df = pd.read_excel(excel_file_path)
print(df.head())

   id  Profitability   debt     RnD  Firm Age  Firm Size  CEO Experience  \
0   1           0.02  23.76   53.09      3.43      21.22               2   
1   2           3.99  41.85   45.30      3.22      21.83               4   
2   3          41.98  61.84   51.16      3.18      20.54               3   
3   4           8.48  25.61  115.49      3.26      20.59               3   
4   5          -1.94  74.24   13.57      3.26      18.73               3   

   foreign sales  CEO education  CEO Gender  marketing  product sentiment  \
0           9.16              2           0   0.801335           0.284925   
1          41.42              1           0   0.385019           0.579495   
2           5.88              3           1   0.839350           0.619687   
3          11.08              2           0   0.643304           0.048441   
4          24.61              1           0   0.747535           0.378619   

   politician on board  POB catagorical  
0             0.158781                

In [17]:
features = ['debt', 'RnD', 'Firm Age', 'Firm Size', 'CEO Experience', 'foreign sales', 'CEO education', 'CEO Gender', 'marketing', 'product sentiment', 'POB catagorical']
X = df[features]
y = df['Profitability']

In [19]:
from sklearn.ensemble import RandomForestRegressor

model = RandomForestRegressor(random_state=42)
model.fit(X_train, y_train)

RandomForestRegressor(random_state=42)

In [20]:
# Get feature importances from the trained model
feature_importances = model.feature_importances_

# Create a pandas Series for better visualization
feature_importance_series = pd.Series(feature_importances, index=X_train.columns)

# Sort the features by importance
sorted_feature_importance = feature_importance_series.sort_values(ascending=False)

# Print the feature importance scores
print("Feature Importance Scores:")
print(sorted_feature_importance)

Feature Importance Scores:
debt                 0.300790
CEO Gender           0.155771
Firm Size            0.142648
RnD                  0.108284
Firm Age             0.060365
CEO Experience       0.053216
marketing            0.052356
foreign sales        0.048258
product sentiment    0.038825
CEO education        0.021732
POB catagorical      0.017755
dtype: float64


In [23]:
from sklearn.linear_model import Lasso
import numpy as np

# Initialize LASSO model (you might need to adjust alpha)
lasso = Lasso(alpha=0.1) # alpha is the regularization strength

# Fit LASSO model
lasso.fit(X, y)

# Get the coefficients
lasso_coefficients = pd.Series(lasso.coef_, index=X.columns)

# Identify selected features (non-zero coefficients)
selected_features_lasso = lasso_coefficients[lasso_coefficients != 0].index.tolist()

print("Features selected by LASSO regularization:")
print(selected_features_lasso)

Features selected by LASSO regularization:
['debt', 'RnD', 'Firm Age', 'Firm Size', 'CEO Experience', 'foreign sales', 'CEO education', 'CEO Gender', 'marketing', 'product sentiment', 'POB catagorical']


## LASSO Coefficients

### Subtask:
Display the coefficients from the trained LASSO model.

**Reasoning**:
Display the calculated LASSO coefficients to see the magnitude and sign of the relationship between each feature and the target variable after regularization.

In [24]:
print("LASSO Coefficients:")
print(lasso_coefficients)

LASSO Coefficients:
debt                 -0.191118
RnD                   0.030952
Firm Age             -2.747879
Firm Size             0.709311
CEO Experience        2.020387
foreign sales        -0.014693
CEO education         2.189313
CEO Gender           10.048181
marketing             3.942607
product sentiment    -1.566340
POB catagorical      -1.725515
dtype: float64


In [25]:
import statsmodels.api as sm

# Add a constant (intercept) to the features
X_with_intercept = sm.add_constant(X)

# Create and fit the OLS model with robust standard errors
model_ols = sm.OLS(y, X_with_intercept)
results_ols = model_ols.fit(cov_type='HC1') # HC1 is a common type of robust standard error

# Print the summary of the regression results
print(results_ols.summary())

                            OLS Regression Results                            
Dep. Variable:          Profitability   R-squared:                       0.293
Model:                            OLS   Adj. R-squared:                  0.222
Method:                 Least Squares   F-statistic:                     3.843
Date:                Wed, 08 Oct 2025   Prob (F-statistic):           0.000107
Time:                        05:27:31   Log-Likelihood:                -464.38
No. Observations:                 122   AIC:                             952.8
Df Residuals:                     110   BIC:                             986.4
Df Model:                          11                                         
Covariance Type:                  HC1                                         
                        coef    std err          z      P>|z|      [0.025      0.975]
-------------------------------------------------------------------------------------
const                -7.2593     15.90

In [26]:
features = ['debt', 'CEO Gender']
X = df[features]
y = df['Profitability']

## Evaluate the model with selected features

### Subtask:
Evaluate the performance of the model trained with selected features using appropriate metrics.

**Reasoning**:
Calculate the MSE and R2 scores for the predictions made with the model trained on selected features.

In [31]:
from sklearn.metrics import mean_squared_error, r2_score

mse_selected = mean_squared_error(y_test, y_pred_selected)
r2_selected = r2_score(y_test, y_pred_selected)

print(f"Mean Squared Error (MSE) with selected features: {mse_selected}")
print(f"R-squared (R2) Score with selected features: {r2_selected}")

Mean Squared Error (MSE) with selected features: 181.3776670989862
R-squared (R2) Score with selected features: -2.939028716334892


## Feature Importance for Model with Selected Features

### Subtask:
Display the feature importance scores from the Random Forest model trained with selected features.

**Reasoning**:
Access the `feature_importances_` attribute of the `model_selected_features` and display the scores along with the corresponding feature names.

In [30]:
# Get feature importances from the trained model with selected features
feature_importances_selected = model_selected_features.feature_importances_

# Create a pandas Series for better visualization
feature_importance_series_selected = pd.Series(feature_importances_selected, index=X_train.columns)

# Sort the features by importance
sorted_feature_importance_selected = feature_importance_series_selected.sort_values(ascending=False)

# Print the feature importance scores
print("Feature Importance Scores (Selected Features):")
print(sorted_feature_importance_selected)

Feature Importance Scores (Selected Features):
debt          0.814748
CEO Gender    0.185252
dtype: float64


## Make predictions with selected features

### Subtask:
Use the trained model with selected features to make predictions on the testing data.

**Reasoning**:
Use the trained Random Forest Regressor model (with selected features) to predict the profitability on the test data.

In [29]:
y_pred_selected = model_selected_features.predict(X_test)

## Train Random Forest model with selected features

### Subtask:
Train a Random Forest Regressor model using the selected features.

**Reasoning**:
Train the Random Forest Regressor model on the training data with the selected features.

In [28]:
from sklearn.ensemble import RandomForestRegressor

model_selected_features = RandomForestRegressor(random_state=42)
model_selected_features.fit(X_train, y_train)

RandomForestRegressor(random_state=42)

## Split the data with selected features

### Subtask:
Split the data with the reduced set of features into training and testing sets using an 80/20 ratio.

**Reasoning**:
Split the data (using the updated X with selected features) into training and testing sets.

In [27]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [32]:
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA
import pandas as pd

# Define the features for PCA
features_for_pca = ['debt', 'RnD', 'Firm Age', 'Firm Size', 'CEO Experience', 'foreign sales', 'CEO education', 'CEO Gender', 'marketing', 'product sentiment', 'POB catagorical']
X_pca = df[features_for_pca]

# Standardize the data before applying PCA
scaler = StandardScaler()
X_pca_scaled = scaler.fit_transform(X_pca)

# Perform PCA
pca = PCA()
X_pca_result = pca.fit_transform(X_pca_scaled)

# Get the loading factors
loading_factors = pd.DataFrame(pca.components_.T, columns=[f'PC{i+1}' for i in range(pca.n_components_)], index=features_for_pca)

# Print the loading factors
print("Principal Component Loading Factors:")
print(loading_factors)

Principal Component Loading Factors:
                        PC1       PC2       PC3       PC4       PC5       PC6  \
debt              -0.174775  0.028798  0.183165 -0.471071  0.582505  0.493400   
RnD                0.299138 -0.243715  0.298815  0.033350  0.318054 -0.350591   
Firm Age           0.382100 -0.084607 -0.094143 -0.410329 -0.402699  0.393862   
Firm Size          0.335501  0.186497 -0.397141 -0.341122  0.104077 -0.202928   
CEO Experience     0.319564 -0.429258 -0.370291  0.011194  0.003411  0.211945   
foreign sales     -0.110608  0.043424 -0.428206  0.522498  0.300299  0.453746   
CEO education      0.411225  0.424441  0.371847  0.015326  0.097420  0.215859   
CEO Gender         0.430913  0.378530  0.101990  0.414389 -0.089332  0.137067   
marketing         -0.133428 -0.275879  0.472301  0.122456 -0.279001  0.320725   
product sentiment -0.192805  0.470148 -0.089772 -0.121111  0.092701 -0.090508   
POB catagorical    0.311904 -0.301833  0.101490  0.132288  0.438673 -0.1